src/main/java/de/tum/in/www1/artemis/service/learningpath/LearningPathRecommendationService.java

```diff
@@ -0,0 +1,597 @@
+        if (weight == 0) {
+            return EXTENDS_OR_ASSUMES_UTILITY;
+        }
+        return (1 / weight) * EXTENDS_OR_ASSUMES_UTILITY;
+
+    }
+
+    /**
+     * Gets the utility of the competency with respect to users mastery progress within the competency.
+     *
+     * @param competency the competency for which the utility should be computed
+     * @param state      the current state of the recommendation system
+     * @return mastery utility of the competency
+     */
+    private static double computeMasteryUtility(Competency competency, RecommendationState state) {
+        return state.competencyMastery.get(competency.getId()) * MASTERY_PROGRESS_UTILITY;
+    }
+
+    /**
+     * Analyzes the current progress within the learning path and generates a recommended ordering of learning objects in a competency.
+     *
+     * @param learningPath the learning path that should be analyzed
+     * @param competency   the competency
+     * @param state        the current state of the recommendation
+     * @return the recommended ordering of learning objects
+     */
+    public List<LearningObject> getRecommendedOrderOfLearningObjects(LearningPath learningPath, Competency competency, RecommendationState state) {
+        var pendingLectureUnits = competency.getLectureUnits().stream().filter(lectureUnit -> !lectureUnit.isCompletedFor(learningPath.getUser())).toList();
+        ArrayList<LearningObject> recommendedOrder = new ArrayList<>(pendingLectureUnits);
+
+        // early return if competency can be trivially mastered
+        if (canBeMasteredWithoutExercises(competency)) {
+            return recommendedOrder;
+        }
+
+        final var combinedPriorConfidence = computeCombinedPriorConfidence(competency, state);
+        final var pendingExercises = competency.getExercises().stream().filter(exercise -> !learningObjectService.isCompletedByUser(exercise, learningPath.getUser()))
+                .collect(Collectors.toSet());
+        final var numberOfExercisesRequiredToMaster = predictNumberOfExercisesRequiredToMaster(learningPath, competency, combinedPriorConfidence, pendingExercises.size());
+        log.warn("numberOfExercisesRequiredToMaster:" + numberOfExercisesRequiredToMaster);
+        HashMap<DifficultyLevel, Set<Exercise>> difficultyLevelMap = generateDifficultyLevelMap(competency.getExercises());
+        if (numberOfExercisesRequiredToMaster >= competency.getExercises().size()) {
+            scheduleAllExercises(recommendedOrder, difficultyLevelMap);
+            return recommendedOrder;
+        }
+        final var recommendedExerciseDistribution = getRecommendedExerciseDistribution(numberOfExercisesRequiredToMaster, combinedPriorConfidence);
+        if (Arrays.stream(recommendedExerciseDistribution).sum() >= competency.getExercises().size()) {
+            scheduleAllExercises(recommendedOrder, difficultyLevelMap);
```

Comment:
Why can this happen now but was not caught in the if before? Because the exercises have the "wrong" difficulty and you predict that you cannot reach the mastery with them?
Maybe explain this a bit in a comment